### PR TITLE
Merge OAuth2 scopes

### PR DESCRIPTION
**Pull request by taojang**
The problem was when merging two swagger doc together, the OAuth2 scopes are not merged together properly because the `Monoid InsOrdMap` has behavior of replacing the value at the same key.

The tactical solution is to make a `SwaggerMonoid` overlapping instance for it. It only merges scopes when `SecurityScheme`
* have same key in security definitions map
* both are `SecuritySchemeOAuth2`
* flows are same

### CODE REVIEW

**Comment by fizruk** [comment]
Makes sense! Thank you for your contribution!

**Comment by phadej** [comment]
Can't we rather change the monoid instance of `SecurityScheme`?

**Comment by taojang** [comment]
@phadej I'm fairly new to this library, I couldn't find the monoid instance of `SecurityScheme`. Have I overlooked something?

**Comment by taojang** [comment]
@phadej I noticed master does not build with stack, so I guess my PR is not going to build on CI after rebasing.

I checked again that there's no instance for `SecurityScheme`. My humble opinion is rather not to have a semigroup instance for it, because:

- it's not directly helping in this situation, because `_swaggerSecurityDefinitions` has type `Definitions SecurityScheme` which is alias to `InsOrdMap Text SecurityScheme`. By default it uses `InsOrdMap.union` as `mappend` which overwrites value with same key
- such instance could give a false impression that it's ok to combine two `SecurityScheme` together. But in reality it only combines scopes together for oauth2 under certain condition. Otherwise it would throw one value away

What do you think? I can do everything just get this merged :grin:

**Comment by phadej** [comment]
I'm not fun of overlapping instances.

@fizruk could we rather have `newtype SecurityDefinitions = SecurityDefinition (Definitions SecurityScheme)`

**Comment by phadej** [comment]
And thinking again, why it's ok to monoidically combine  `Definitions SecurityScheme`, but not single `SecurityScheme`. Can @taojang open that up?

**Comment by taojang** [comment]
@phadej thx for pointing out the direction. I updated the implementation according to your suggestion

File: src/Data/Swagger/Internal.hs [hunk 1]
```diff
@@ -1104,6 +1120,9 @@ instance ToJSON PathItem where
 instance ToJSON Example where
   toJSON = toJSON . Map.mapKeys show . getExample
 
+instance ToJSON SecurityDefinitions where
```
Comment (reactormonk) [inline src/Data/Swagger/Internal.hs]: You can derive these with `GeneralizedNewtypeDeriving`.

**Comment by fizruk** [comment]
@phadej are you happy with this change? Should I merge?

**Comment by i-am-the-slime** [comment]
Is there any chance this will be merged one day?

**Comment by fisx** [comment]
I've just received commit authority, and I will merge this if (1) nobody objects in time; and (2) you resolve the merge conflict.

Thanks!

**Comment by i-am-the-slime** [comment]
@taojang

**Comment by taojang** [comment]
Thanks for the heads-up! I will resolve it before next week :)

**Comment by taojang** [comment]
@fisx I have resolved the merge conflict

File: src/Data/Swagger/Internal.hs [hunk 2]
```diff
@@ -904,6 +920,17 @@ instance Monoid Example where
   mempty = genericMempty
   mappend = (<>)
 
+instance Semigroup SecurityScheme where
+  (<>) = mergeSecurityScheme
+
+instance Semigroup SecurityDefinitions where
+  (SecurityDefinitions sd1) <> (SecurityDefinitions sd2) =
+     SecurityDefinitions $ InsOrdHashMap.unionWith (<>) sd1 sd2
```
Comment (fisx) [inline src/Data/Swagger/Internal.hs]: (Shouldn't there be an `instance Semigroup InsOrdHashMap`?  Same for `Monoid`?)

File: src/Data/Swagger/Internal.hs [hunk 3]
```diff
@@ -904,6 +920,17 @@ instance Monoid Example where
...
+
+instance Monoid SecurityDefinitions where
+  mempty = SecurityDefinitions $ InsOrdHashMap.empty
```
Comment (fisx) [inline src/Data/Swagger/Internal.hs]:
```suggestion
  mempty = SecurityDefinitions InsOrdHashMap.empty
```

**Pull request merged**